sys/net/link_layer/ng_nomac/ng_nomac.c

```diff
@@ -0,0 +1,155 @@
+
+/**
+ * @{
+ * @ingroup     net_nomac
+ * @file
+ * @brief       Implementation of the NOMAC MAC protocol
+ *
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @}
+ */
+
+#include <errno.h>
+
+#include "kernel.h"
+#include "msg.h"
+#include "thread.h"
+#include "net/ng_nomac.h"
+#include "net/ng_netreg.h"
+#include "net/ng_pkt.h"
+#include "net/ng_nettype.h"
+#include "net/ng_netdev.h"
+#include "net/ng_netapi.h"
+#include "net/ng_netif.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+/**
+ * @brief   Function called by the device driver on device events
+ *
+ * @param[in] event         type of event
+ * @param[in] data          optional parameter
+ */
+static void _event_cb(ng_netdev_event_t event, void *data)
+{
+    DEBUG("nomac: event triggered -> %i\n", event);
+    /* NOMAC only understands the RX_COMPLETE event... */
+    if (event == NETDEV_EVENT_RX_COMPLETE) {
+        ng_pktsnip_t *pkt;
+        ng_netreg_entry_t *sendto;
+
+        /* get pointer to the received packet */
+        pkt = (ng_pktsnip_t *)data;
+        /* find out, who to send the packet to */
+        sendto = ng_netreg_lookup(pkt->next->type, 0);
+
+#if ENABLE_DEBUG
+        if (sendto == NULL) {
+            DEBUG("nomac: unable to forward packet of type %i\n",
+                  pkt->next->type);
+        }
+#endif
+
+        /* send the packet to everyone interested in it's type */
+        while (sendto != NULL) {
+            DEBUG("nomac: sending pkt %p to PID %u\n", pkt, sendto->pid);
+            ng_netapi_send(sendto->pid, pkt);
+            sendto = ng_netreg_getnext(sendto);
+        }
+    }
+}
+
+/**
+ * @brief   Startup code and event loop of the NOMAC layer
+ *
+ * @param[in] args          expects a pointer to the underlying netdev device
+ *
+ * @return                  never returns
+ */
+static void *_nomac_thread(void *args)
+{
+    ng_netdev_t *dev = (ng_netdev_t *)args;
+    ng_netapi_opt_t *opt;
+    int res;
+    msg_t msg, reply, msg_queue[NG_NOMAC_MSG_QUEUE_SIZE];
+
+    /* setup the MAC layers message queue */
+    msg_init_queue(msg_queue, NG_NOMAC_MSG_QUEUE_SIZE);
+    /* save the PID to the device descriptor and register the device */
+    dev->mac_pid = thread_getpid();
+    ng_netif_add(dev->mac_pid);
+    /* register the event callback with the device driver */
+    dev->driver->add_event_callback(dev, _event_cb);
+
+    /* start the event loop */
+    while (1) {
+        DEBUG("nomac: waiting for incoming messages\n");
+        msg_receive(&msg);
+        /* dispatch NETDEV and NETAPI messages */
+        switch (msg.type) {
+            case NG_NETDEV_MSG_TYPE_EVENT:
+                DEBUG("nomac: NG_NETDEV_MSG_TYPE_EVENT received\n");
+                dev->driver->isr_event(dev, msg.content.value);
+                break;
+            case NG_NETAPI_MSG_TYPE_SND:
+                DEBUG("nomac: NG_NETAPI_MSG_TYPE_SND received\n");
+                dev->driver->send_data(dev, (ng_pktsnip_t *)msg.content.ptr);
```

Comment:
As @gebart pointed out in https://github.com/RIOT-OS/RIOT/pull/2426/files#r26117750: This can still be `NULL`. Maybe the device does not implement `send_data` (or any other of the methods).